File: agents/numerology_engine.py
import math
from datetime import datetime
from typing import Dict, List, Optional

class NumerologyEngine:
    """
    Core engine for numerological calculations.
    Supports Pythagorean system for names and standard digit reduction.
    """
    
    PYTHAGOREAN_CHART = {
        'A': 1, 'J': 1, 'S': 1,
        'B': 2, 'K': 2, 'T': 2,
        'C': 3, 'L': 3, 'U': 3,
        'D': 4, 'M': 4, 'V': 4,
        'E': 5, 'N': 5, 'W': 5,
        'F': 6, 'O': 6, 'X': 6,
        'G': 7, 'P': 7, 'Y': 7,
        'H': 8, 'Q': 8, 'Z': 8,
        'I': 9, 'R': 9
    }

    @staticmethod
    def reduce_digits(n: int, reduce_to_master: bool = True) -> int:
        """Reduces a number to a single digit or master number (11, 22, 33)."""
        while n > 9:
            if reduce_to_master and n in [11, 22, 33]:
                return n
            n = sum(int(digit) for digit in str(n))
        return n

    def calculate_life_path(self, dob: str) -> int:
        """
        Calculates Life Path Number from date of birth (YYYY-MM-DD).
        """
        try:
            date_obj = datetime.strptime(dob, "%Y-%m-%d")
            year_red = self.reduce_digits(date_obj.year)
            month_red = self.reduce_digits(date_obj.month)
            day_red = self.reduce_digits(date_obj.day)
            
            return self.reduce_digits(year_red + month_red + day_red)
        except ValueError:
            raise ValueError("Date must be in YYYY-MM-DD format.")

    def calculate_expression(self, name: str) -> int:
        """
        Calculates Expression (Destiny) Number from full name.
        """
        name = name.upper().replace(" ", "")
        total = sum(self.PYTHAGOREAN_CHART.get(char, 0) for char in name)
        return self.reduce_digits(total)

    def calculate_personal_year(self, dob: str, year: Optional[int] = None) -> int:
        """
        Calculates Personal Year Number for a given year (defaults to current).
        """
        if year is None:
            year = datetime.now().year
            
        try:
            date_obj = datetime.strptime(dob, "%Y-%m-%d")
            month_red = self.reduce_digits(date_obj.month)
            day_red = self.reduce_digits(date_obj.day)
            year_red = self.reduce_digits(year)
            
            return self.reduce_digits(month_red + day_red + year_red)
        except ValueError:
            raise ValueError("Date must be in YYYY-MM-DD format.")

    def get_daily_vibration(self, dob: str, date: Optional[str] = None) -> int:
        """
        Calculates the vibration for a specific day.
        """
        if date is None:
            date_obj = datetime.now()
        else:
            date_obj = datetime.strptime(date, "%Y-%m-%d")
            
        personal_year = self.calculate_personal_year(dob, date_obj.year)
        month_red = self.reduce_digits(date_obj.month)
        day_red = self.reduce_digits(date_obj.day)
        
        return self.reduce_digits(personal_year + month_red + day_red)

if __name__ == "__main__":
    # Quick test
    engine = NumerologyEngine()
    print(f"Life Path (1990-01-01): {engine.calculate_life_path('1990-01-01')}")
    print(f"Expression (John Doe): {engine.calculate_expression('John Doe')}")
    print(f"Personal Year 2024: {engine.calculate_personal_year('1990-01-01', 2024)}")

File: frontend/src/utils/numerologyEngine.js

export class NumerologyEngine {
    /**
     * Reduces a number to a single digit or master number (11, 22, 33).
     * @param {number} n 
     * @param {boolean} reduceToMaster 
     * @returns {number}
     */
    static reduceDigits(n, reduceToMaster = true) {
        while (n > 9) {
            if (reduceToMaster && [11, 22, 33].includes(n)) {
                return n;
            }
            n = String(n).split('').reduce((sum, digit) => sum + parseInt(digit), 0);
        }
        return n;
    }

    /**
     * Calculates the Life Path Number.
     * @param {string} dob - Date of birth in YYYY-MM-DD format
     * @returns {number}
     */
    static calculateLifePath(dob) {
        if (!dob) return 0;
        const [year, month, day] = dob.split('-').map(Number);

        // Pythagorean method: Reduce Month, Day, Year separately first
        const rMonth = this.reduceDigits(month);
        const rDay = this.reduceDigits(day);
        const rYear = this.reduceDigits(year);

        return this.reduceDigits(rMonth + rDay + rYear);
    }

    /**
     * Calculates the Personal Year Number.
     * @param {string} dob - Date of birth in YYYY-MM-DD
     * @param {number} targetYear - The year to calculate for
     * @returns {number}
     */
    static calculatePersonalYear(dob, targetYear = new Date().getFullYear()) {
        if (!dob) return 0;
        const [_, month, day] = dob.split('-').map(Number);

        const rMonth = this.reduceDigits(month);
        const rDay = this.reduceDigits(day);
        const rYear = this.reduceDigits(targetYear);

        return this.reduceDigits(rMonth + rDay + rYear);
    }

    /**
     * Calculates the Personal Daily Vibration.
     * @param {string} dob - Date of birth
     * @param {string} targetDate - Date in YYYY-MM-DD (defaults to today)
     * @returns {number}
     */
    static calculateDailyVibration(dob, targetDate) {
        if (!dob) return 0;
        const dateObj = targetDate ? new Date(targetDate) : new Date();
        const personalYear = this.calculatePersonalYear(dob, dateObj.getFullYear());

        // Month + Day + Personal Year
        const month = dateObj.getMonth() + 1;
        const day = dateObj.getDate();

        return this.reduceDigits(month + day + personalYear);
    }

    /**
     * Calculates the Universal Day Number.
     * @param {string} targetDate - Date in YYYY-MM-DD
     * @returns {number}
     */
    static calculateUniversalDay(targetDate) {
        const dateObj = targetDate ? new Date(targetDate) : new Date();
        const year = dateObj.getFullYear();
        const month = dateObj.getMonth() + 1;
        const day = dateObj.getDate();

        // Universal Day = Sum of all digits of the current date
        return this.reduceDigits(year + month + day);
    }

    /**
     * Calculates the Personal Month Number.
     * @param {string} dob - Date of birth
     * @param {number} targetYear 
     * @param {number} targetMonth 
     * @returns {number}
     */
    static calculatePersonalMonth(dob, targetYear, targetMonth) {
        if (!dob) return 0;
        const personalYear = this.calculatePersonalYear(dob, targetYear);
        return this.reduceDigits(personalYear + targetMonth);
    }

    /**
     * Calculates the Personal Week Number.
     * Logic: Personal Month + Calendar Week Number (approximate).
     * @param {string} dob 
     * @param {number} targetYear 
     * @param {number} targetMonth 
     * @param {number} weekNumber - 1-52 (or 1-5 for month)
     * @returns {number}
     */
    static calculatePersonalWeek(dob, targetYear, targetMonth, weekNumber) {
        if (!dob) return 0;
        const personalMonth = this.calculatePersonalMonth(dob, targetYear, targetMonth);
        return this.reduceDigits(personalMonth + weekNumber);
    }
}
